bintray-tidy.sh

```diff
@@ -0,0 +1,256 @@
+cat <<EOF
+                            ~~~ bintray-tidy.sh ~~~
+
+bintray-tidy.sh - delete older versions from packages on Bintray to save space.
+
+USAGE:    ./bintray-tidy.sh [OPTIONS] STRATEGY [[OWNER/]REPO/]PACKAGE
+
+STRATEGIES:
+  max-versions <num>    Keep <num> newest versions and delete the rest.
+  max-days <num>        Delete all versions created more than <num> days ago.
+  archive               Keep some older versions for archival purposes.
+
+OPTIONS:
+  -h, -?, --help        Show this help.
+  -s, --simulate        Show what would be deleted without actually deleting it.
+
+ENVIRONMENT:
+  * BINTRAY_USER, BINTRAY_API_KEY - required
+  * BINTRAY_REPO, BINTRAY_REPO_OWNER - optional
+
+EXAMPLES:
+
+ 1) Keep the 10 most recent versions and delete everything else:
+
+    $  ./bintray-tidy.sh  max-versions  10  MyRepo/CoolPackage
+
+ 2) Simulate running the archive strategy to see what would be deleted:
+
+    $  ./bintray-tidy.sh  -s  archive  DreamTeam/TeamRepo/SharedPackage
+
+                            ~~~ bintray-tidy.sh ~~~
+EOF
+}
+
+main() {
+  initialize_variables
+
+  echo "$0"
+  echo "Tidying package '${BINTRAY_REPO_OWNER}/${BINTRAY_REPO}/${package}'."
+
+  get_remote_versions
+
+  [ "$arg_s" ] && echo "SIMULATION: No versions will actually be deleted from the server." || true
+
+  case "${strategy}" in
+    max-versions )
+      max_versions
+      ;;
+    max-days )
+      max_days
+      ;;
+    archive )
+      archive
+      ;;
+    * )
+      fatal_error "invalid strategy '${strategy}'"
+      exit 1
+      ;;
+  esac
+
+  [ "$arg_s" ] && echo "SIMULATION: Nothing was deleted. Rerun without '-s' to actually delete versions." || true
+}
+
+function initialize_variables() {
+  # Constants
+  API="https://api.bintray.com"
+
+  # Required environment variables
+  BINTRAY_USER="${BINTRAY_USER:?Environment variable missing/empty!}" # env
+  BINTRAY_API_KEY="${BINTRAY_API_KEY:?Environment variable missing/empty!}" # env
+
+  # Optional environment variables
+  BINTRAY_REPO_OWNER="${BINTRAY_REPO_OWNER:-$BINTRAY_USER}" # env, or use BINTRAY_USER
+  BINTRAY_REPO="${BINTRAY_REPO:?Not specified and not set in environment.}" # env or CLI
+
+  # Variables
+  PCK_URL="${API}/packages/${BINTRAY_REPO_OWNER}/${BINTRAY_REPO}/${package}"
+  CURL="curl -u${BINTRAY_USER}:${BINTRAY_API_KEY} \
+         -H Content-Type:application/json \
+         -H Accept:application/json"
+}
+
+function get_remote_versions() {
+  readarray -t REMOTE_VERSIONS < <( \
+  curl -X GET "${PCK_URL}" \
+    | sed -nr 's|.*"versions":\["([^]]*)"\].*|\1|p' \
+    | sed 's|","|\n|g' )
+  echo "$0: ${#REMOTE_VERSIONS[@]} versions found on the remote server."
+}
+
+function max_versions() {
+  # Enforce upper limit on the number of versions:
+  for ((i=0;i<${#REMOTE_VERSIONS[@]};i++)); do
+    if [ "$i" -lt "$max_versions" ]; then
+      print_action "Keeping" "${REMOTE_VERSIONS[$i]}"
+    else
+      delete_version "${REMOTE_VERSIONS[$i]}" "$created"
```

Comment:
@probonopd Just noticed a (harmless) mistake: you should remove "$created" from the end of this line because packages are not checked for creation dates when the "max-versions" strategy is being used, but it is correct on the other lines. It doesn't do any harm here since it will just be empty, but technically it shouldn't be there.